src/virtio/virtqueue.c

```diff
@@ -219,9 +233,11 @@ status virtqueue_alloc(vtdev dev,
 {
     u64 vq_alloc_size = sizeof(struct virtqueue) + size * sizeof(vqmsg);
     virtqueue vq = allocate_zero(dev->general, vq_alloc_size);
-    vq->avail_offset = size * sizeof(struct vring_desc);
-    vq->used_offset = pad(vq->avail_offset + sizeof(*vq->avail) + sizeof(vq->avail->ring[0]) * size, align);
-    bytes alloc = vq->used_offset + pad(sizeof(*vq->used) + sizeof(vq->used->ring[0]) * size, align);
+    bytes avail_offset = size * sizeof(struct vring_desc);
+    bytes used_offset = pad(avail_offset + sizeof(*vq->avail) + sizeof(vq->avail->ring[0]) * size +
+                            sizeof(u16) /* used_event */, align);
```

Comment:
Is the extra 2 bytes for the used_event still appropriate here (and again below) if VIRTIO_F_RING_EVENT_IDX has not been negotiated with the device?

Reply:
While those extra bytes are unused if VIRTIO_F_RING_EVENT_IDX has not been negotiated, they only affect the offset and allocation size of the virtqueue available and used rings. For non-legacy devices, the offsets don't really matter, since the location of the rings is passed to device during configuration (e..g. via VTPCI_R_QUEUE_AVAIL and VTPCI_R_QUEUE_USED for PCI devices). For legacy devices, that require a fixed virtqueue layout, the offset do matter, but as far as I can see the calculation of the offsets was wrong before (and is correct with these changes): the virtio spec (version 1.1), at section 2.6.2 "Legacy Interfaces: A Note on Virtqueue Layout", says that the total virtqueue size is calculated as `ALIGN(sizeof(struct virtq_desc)*qsz + sizeof(u16)*(3 + qsz)) + ALIGN(sizeof(u16)*3 + sizeof(struct virtq_used_elem)*qsz)`. The `3 * sizeof(u16)` part does not match the size of struct vring_avail and struct vring_used in the kernel. By adding another 2 bytes we match the formula. In practice, in the majority of cases this doesn't make a difference, due to the padding (necessary for alignment) inserted between the available and used rings. I guess we never came across a legacy device for which those 2 bytes would change the actual offset between available and used rings, so we never noticed the issue before.

Reply:
Thanks, great explanation!